packages/world/src/systems/NatureSystem.sol

```diff
@@ -33,40 +45,162 @@ contract NatureSystem is System {
+    require(objectType.isGrowable(), "Not growable");
+    require(SeedGrowth._getFullyGrownAt(seed) <= block.timestamp, "Seed cannot be grown yet");
+
+    // When a seed grows, it's removed from circulation
+    // We only update ResourceCount since seeds don't participate in respawning (no need to track positions
+    uint256 seedCount = ResourceCount._get(objectType);
+    // This should never happen if there are seeds in the world obtained from drops
+    require(seedCount > 0, "Not enough seeds in circulation");
+
+    if (objectType.isSeed()) {
+      // Turn wet farmland to regular farmland if mining a seed or crop
+      (EntityId below, ObjectTypeId belowType) = getOrCreateEntityAt(coord - vec3(0, 1, 0));
+      // Sanity check
+      if (belowType == ObjectTypes.WetFarmland) {
+        ObjectType._set(below, ObjectTypes.Farmland);
+      }
+
+      ObjectType._set(seed, objectType.getCrop());
+    } else if (objectType.isSapling()) {
+      // Grow the tree (replace the seed with the trunk and add blocks)
+      TreeData memory treeData = TreeLib.getTreeData(objectType);
+
+      (uint32 trunkHeight, uint32 leaves) = _growTree(seed, coord, treeData, objectType);
+
+      uint128 seedEnergy = ObjectTypeMetadata._getEnergy(objectType);
+      uint128 trunkEnergy = trunkHeight * ObjectTypeMetadata._getEnergy(treeData.logType);
+      uint128 leafEnergy = leaves * ObjectTypeMetadata._getEnergy(treeData.leafType);
+
+      uint128 energyToReturn = seedEnergy - trunkEnergy - leafEnergy;
+
+      if (energyToReturn > 0) {
+        addEnergyToLocalPool(coord, energyToReturn);
+      }
+    }
+  }
+
+  function _growTree(EntityId seed, Vec3 baseCoord, TreeData memory treeData, ObjectTypeId saplingType)
+    private
+    returns (uint32, uint32)
+  {
+    uint32 trunkHeight = _growTreeTrunk(seed, baseCoord, treeData);
+
+    if (trunkHeight <= 2) {
+      // Very small tree, no leaves
+      return (trunkHeight, 0);
+    }
+
+    // Adjust if the tree is blocked
+    if (trunkHeight < treeData.trunkHeight) {
+      trunkHeight = trunkHeight + 1; // Still allow one layer above the trunk
+    }
+
+    (Vec3[] memory fixedLeaves, Vec3[] memory randomLeaves) = TreeLib.getLeafCoords(saplingType);
+
+    // Initial seed for randomness
+    uint256 rand = uint256(keccak256(abi.encodePacked(block.timestamp, baseCoord)));
+
+    uint32 leafCount;
+    ObjectTypeId leafType = treeData.leafType;
+
+    for (uint256 i = 0; i < fixedLeaves.length; ++i) {
+      Vec3 rel = fixedLeaves[i];
+      if (rel.y() > int32(trunkHeight)) {
+        break;
+      }
+
+      if (_tryCreateLeaf(leafType, baseCoord + rel)) {
+        ++leafCount;
+      }
+    }
+
+    for (uint256 j = 0; j < randomLeaves.length; ++j) {
+      Vec3 rel = randomLeaves[j];
+      if (rel.y() > int32(trunkHeight)) {
+        break;
+      }
+
+      rand = uint256(keccak256(abi.encodePacked(rand, j))); // evolve RNG
+
+      if (rand % 100 < 40) continue; // 40 % trimmed
```

Comment:
what is this work?

Reply:
the original tree algo applied 40% chance of creating a leaf for certain leafs like corners